tensorflow/c/eager/tape.h

```diff
@@ -1062,7 +1105,73 @@ Status ForwardAccumulator<Gradient, BackwardFunction, TapeTensor>::Accumulate(
         output_tensors, backward_function_getter, backward_function_deleter,
         in_grads, &forward_grads));
   } else {
-    TF_RETURN_IF_ERROR((*forward_function)(in_grads, &forward_grads));
+    TF_RETURN_IF_ERROR((*forward_function)(in_grads, &forward_grads, false));
+  }
+  for (int i = 0; i < forward_grads.size(); ++i) {
+    if (forward_grads[i] != nullptr) {
+      int64 tensor_id = output_tensors[i].GetID();
+      auto existing = accumulated_gradients_.find(tensor_id);
+      if (existing != accumulated_gradients_.end()) {
+        // This is a somewhat odd case to be in, since it means we have two
+        // operations which supposedly both created the same Tensor. It comes up
+        // in recompute_grad, where the gradients have the same value. However,
+        // only the original gradient is connected to everything else, so we
+        // should still use that.
+        vspace_.DeleteGradient(forward_grads[i]);
+      } else {
+        accumulated_gradients_[output_tensors[i].GetID()] = forward_grads[i];
+      }
+    }
+  }
+  return Status::OK();
+}
+
+template <typename Gradient, typename BackwardFunction, typename TapeTensor>
+Status
+ForwardBatchAccumulator<Gradient, BackwardFunction, TapeTensor>::Accumulate(
```

Comment:
One thing that seems plausible is to just keep the boolean around, then we can add broadcasting for un-batched tangents in Python. Does that work, and do we need different C++ classes then?